### DISCUSSION TITLE
Upcoming 2.0 release

### DISCUSSION THREAD

**Discussion by donalffons** [donalffons/opencascade.js#16] [General]
Here are some notes on the upcoming release of the library.

:package: **Web-Assembly Modules**

- The library will be split up into several smaller WASM-modules, as this gives the following advantages:
  - For client side / browser applications, it will be much easier to pick and choose the parts of the library that are required for a project. This will greatly reduce the size of the application.
  - Startup times will be reduced, as less WASM code has to be compiled and instantiated. This is especially relevant for slow hardware, old browsers and older versions of NodeJS.
- There will be two "flavors" of the library. Split up into "Modules" (not to be confused with WASM-modules) and "Packages" - both of which are OpenCascade terminology.
  - Packages are the smallest building block of the OpenCascade library. There are roughly 300 packages in total.
  - Several packages are combined into modules. There are roughly 50 modules in total.

:mag: **Typescript definitions**

Typescript definitions will be generated for all bindings and some parts of the Emscripten library (like the `FS` object for filesystem manipulations). This will give "Intellisense" support in your favorite IDE and will allow static type checking to some degree.

:rocket: **Updated OCCT**

The newly released version 7.5.0 of OpenCascade will be supported.

:beetle: **Debugging**

There will be debug and release builds of the library, which might make it easier to debug some of the more challenging issues.

:computer: **NodeJS**

Support for NodeJS will be explicitly tested.

:gear: **Build System**

The build system has been rewritten in large parts with the following major improvements:
  - Support for multi-threaded builds. As Emscripten is mostly single-threaded, this will greatly reduce the time to build.
  - Typedef'd template classes will be supported by the build system, which will greatly increase the number of usable APIs of the library.
  - The new build system will most likely resolve some of the current issues with the build system (docker containers becoming zombies). Therefore, proper CI/CD will most likely be coming back to this library.

Feedback and questions are always welcome! :v:

**Comment by womaas** [comment]
Hi.
Now the dist **folder** does not has a wasm module. When are you planning to add it? Is it possible to generate wasm yourself from your sources? If so, how?

**Reply by donalffons** [reply]
Hey,
thanks for reporting that. After an Emscripten upgrade, some builds were failing and since there is currently no testing for the `beta` builds, some broken packages were accidentally released.
Please try with the latest build and let me know if this solves your problem. Also, in case you were importing "module.TKernel.wasm" anywhere in your code, please remove this statement (it is not included by default in "opencascade.js/wasm".

**Comment by womaas** [comment]
"Please try with the latest build"
Which build is the latest? Where can I download it?

пн, 10 мая 2021 г. в 09:53, Sebastian Alff ***@***.***>:

**Reply by donalffons** [reply]
```bash
npm install opencascade.js@beta
```
as mentioned in the readme here on Github. Please be aware that the beta builds will most likely encounter breaking changes in the future.

**Comment by womaas** [comment]
Thank you. What do you recommend to use, mainly for modeling: the latest
beta build or the stable 1.1 release?

пн, 10 мая 2021 г. в 14:29, Sebastian Alff ***@***.***>:

**Reply by donalffons** [reply]
I would recommend to use the beta builds, since they are the most complete version of the library and they are using the most recent OpenCascade version 7.5.2. However, there are still several missing parts and loose ends, which I hope I'm able to fix eventually.

**Comment by womaas** [comment]
Thanks!

пн, 10 мая 2021 г. в 20:44, Sebastian Alff ***@***.***>:

**Comment by womaas** [comment]
Hi. How to use @beta in browser without bundler (as ES6
module)?

пн, 10 мая 2021 г. в 21:59, Viktor Viktor ***@***.***>:

**Reply by donalffons** [reply]
I personally haven't tested it without a bundler, yet. The "raw" emscripten outputs are located under `opencascade.js/dist/opencascade.full.[js/wasm]`. You might be able to use those files directly.

**Comment by womaas** [comment]
I have changed

initOpenCascade(...mainWasm= "full path to opencascade.full.wasm"...)

And everything loads as it should, but the initialization process itself is
very slow. OC becomes available in about 20 seconds. Is that how it should
be?

**Reply by donalffons** [reply]
Apologies for the late reply. The long instantiation time is expected - unfortunately. It is mentioned in the docs [here](https://github.com/donalffons/opencascade.js/tree/master/doc#workflow-when-building-apps). The main reason for the slowness is Emscripten's process of "crafting" JavaScript functions for every binding when the module is loaded. I would recommend that you create a custom build for your production app, which will load significantly faster.
